src/tests/Interop/COM/ComWrappers/API/Program.cs

```diff
@@ -962,6 +963,74 @@ public void ValidateAggregationWithReferenceTrackerObject()
 
             Assert.Equal(0, allocTracker.GetCount());
         }
+
+        [Fact]
+        public void ComWrappersNoLockAroundQueryInterface()
+        {
+            Console.WriteLine($"Running {nameof(ComWrappersNoLockAroundQueryInterface)}...");
+
+            var cw = new RecursiveSimpleComWrappers();
+
+            IntPtr comObject = cw.GetOrCreateComInterfaceForObject(new RecursiveCrossThreadQI(cw), CreateComInterfaceFlags.None);
+            try
+            {
+                _ = cw.GetOrCreateObjectForComInstance(comObject, CreateObjectFlags.TrackerObject);
+            }
+            finally
+            {
+                Marshal.Release(comObject);
+            }
+        }
+
+        private class RecursiveCrossThreadQI(ComWrappers? wrappers) : ICustomQueryInterface
+        {
+            CustomQueryInterfaceResult ICustomQueryInterface.GetInterface(ref Guid iid, out IntPtr ppv)
+            {
+                ppv = IntPtr.Zero;
+                if (iid == ComWrappersHelper.IID_IReferenceTracker && wrappers is not null)
+                {
+                    Console.WriteLine("Attempting to create a new COM object on a different thread.");
```

Comment:
Is the issue here thread affinity or COM apartment affinity?

Reply:
I believe the original deadlock was due to COM apartment affinity with the STA vs a Free-threaded apartment thread.

In any case, this test ensures that we're able to create an RCW while we're in the middle of creating an RCW on another thread, which was what was occurring in the original issue.